### PR TITLE
Remove InPortData and OutPortData

### PR DESCRIPTION
**Pull request by ikeough**
### Purpose

This PR represents a substantial refactoring of ports and connectors.

This PR removes the `InPortData` and `OutPortData` collections on `NodeModel`. These collections pre-date the `InPorts` and `OutPorts` collections, and were previously deprecated. They acted as a way to store initialization data for ports during the construction of the node, with the assumption that at some point after they were filled, `NodeModel.RegisterAllPorts` would be called, subsequently creating/updating ports on the nodes. This indirection was not obvious or useful and led to the `InportData` and `OutportData` collections getting out of sync with the `InPorts` and `OutPorts` collections. 

In the beginning, collections in Dynamo were not observable, so we created a lot of events like `NodeModel.PortConnected` and `NodeModel.ConnectorConnected` when ports and connectors were added. As we have moved to using observable collections, it's much easier to deal with the creation and deletion of ports, connectors, nodes, etc., by simply handling the collection changed events for those collections and doing the right thing. This PR gets us part of the way towards a fully observable methodology by adjusting `NodeModel` to watch the collection changed events for the `InPorts` and `OutPorts` collections, and consolidating logic previously spread among numerous event handlers into those collection change handlers.

Additionally, this PR removes a number of methods which seemed like a good idea three years ago, but now only add confusion. For example. `PortModel.Connect` was a very thin wrapper around `PortModel.Connectors.Add(someNewConnector)`. These thin wrappers have been removed except in cases where the wrapper contained some assertion logic.

### Declarations

Check these if you believe they are true
- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI. (In progress)
- [x] Snapshot of UI changes, if any.
- [x] Follows [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) standards (_No change of public methods or types etc.._)
### Reviewers

@mjkkirschner

### CODE REVIEW

**Comment by ramramps** [comment]
@ikeough DynamoStorage is using node.inportdata.
image

**Comment by ikeough** [comment]
@ramramps Thanks! I'll fix that.

File: src/DynamoCore/Graph/Nodes/NodeModel.cs [hunk 1]
```diff
@@ -895,6 +876,15 @@ protected NodeModel()
...
             RaisesModificationEvents = true;
+
+            InPorts.CollectionChanged += PortsCollectionChanged;
+            OutPorts.CollectionChanged += PortsCollectionChanged;
+        }
+
+        private void PortsCollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e)
+        {
+            ValidateConnections();
+            ConfigureSnapEdges(sender == InPorts?InPorts : OutPorts);
```
Comment (mjkkirschner) [inline src/DynamoCore/Graph/Nodes/NodeModel.cs]: sender will be a collection? not the modified element?

File: src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs [hunk 2]
```diff
@@ -268,8 +268,8 @@ void _port_PropertyChanged(object sender, System.ComponentModel.PropertyChangedE
                 case "Center":
                     RaisePropertyChanged("Center");
                     break;
-                case "DefaultValueEnabled":
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs]: what was the reason to get rid of this?

File: src/Libraries/CoreNodeModels/HigherOrder/MapCombineLacing.cs [hunk 3]
```diff
@@ -263,11 +263,11 @@ public class Reduce : VariableInputNode
 
         public Reduce()
         {
-            InPortData.Add(new PortData("reductor", Resources.ReducePortDataReductorToolTip));
-            InPortData.Add(new PortData("seed", Resources.ReducePortDataSeedToolTip));
-            InPortData.Add(new PortData("list1", Resources.PortDataListToolTip + " #1"));
+            InPorts.Add(new PortModel(PortType.Input, this, new PortData("reductor", Resources.ReducePortDataReductorToolTip)));
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodeModels/HigherOrder/MapCombineLacing.cs]: just curious... what happens when a zero touch node author adds a PortModel of type Output to the `InPorts`collection... do we crash?

**Comment by mjkkirschner** [comment]
@ikeough this looks good to me. I have a few questions though, see some of the comments, also, does the effectively break all `nodeModel`nodes made by third parties, or any other others up until now. For example, the `Dynamo Samples` repo will need to be updated and `Revit` UI nodes?

Is there any work necessary for zero touch node authors?

File: src/DynamoCore/Graph/Nodes/NodeModel.cs [hunk 4]
```diff
@@ -895,6 +876,15 @@ protected NodeModel()
             ArgumentLacing = LacingStrategy.Disabled;
 
             RaisesModificationEvents = true;
+
+            InPorts.CollectionChanged += PortsCollectionChanged;
```
Comment (ikeough) [inline src/DynamoCore/Graph/Nodes/NodeModel.cs]: Ah yes. Nice catch @mjkkirschner.

File: src/DynamoCore/Graph/Nodes/NodeModel.cs (hunk 1 above)
Comment (ikeough) [inline src/DynamoCore/Graph/Nodes/NodeModel.cs]: Yes. The `CollectionChanged` event's `sender` is the collection. You can get at the elements that have been changed by using the [`NotifyCollectionChangedEventArgs`](https://msdn.microsoft.com/en-us/library/system.collections.specialized.notifycollectionchangedeventargs(v=vs.110).aspx).

File: src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs (hunk 2 above)
Comment (ikeough) [inline src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs]: `DefaultValueEnabled` was removed on the `NodeModel`, so the change event can be removed here. The property was removed from `NodeModel` in favor of using the more precise `DefaultValue == null`.

File: src/Libraries/CoreNodeModels/HigherOrder/MapCombineLacing.cs (hunk 3 above)
Comment (ikeough) [inline src/Libraries/CoreNodeModels/HigherOrder/MapCombineLacing.cs]: Nothing. This design is not great and needs to be cleaned up, but I didn't want to do it as part of this PR. We should remove `PortType` because it doesn't have any real meaning. Whether a port is an "input" or an "output" can be determined by whether it's in the `InPorts` or the `OutPorts` collections respectively.

**Comment by ikeough** [comment]
@mjkkirschner Yes. This will break third party nodes, the samples, and Revit. That is why this is a 2.0 change. Fortunately, the `InportData` and `OutportData` collections have been marked as deprecated for several versions.

File: src/DynamoCore/Core/CustomNodeDefinition.cs [hunk 5]
```diff
@@ -50,7 +50,7 @@ public class CustomNodeDefinition : IFunctionDescriptor
             if (outputs.Any())
             {
                 topMost.AddRange(
-                    outputs.Where(x => x.HasInput(0)).Select(x => Tuple.Create(0, x as NodeModel)));
+                    outputs.Where(x => x.InPorts[0].IsConnected).Select(x => Tuple.Create(0, x as NodeModel)));
```
Comment (ikeough) [inline src/DynamoCore/Core/CustomNodeDefinition.cs]: API Change: `NodeModel.HasInput(...)` has been removed. Use `PortModel.IsConnected` instead.

File: src/DynamoCore/Core/CustomNodeManager.cs [hunk 6]
```diff
@@ -767,7 +771,7 @@ internal bool TryGetNodeInfo(string name, out CustomNodeInfo info)
                         selectedNodeSet.SelectMany(
                             node =>
                                 Enumerable.Range(0, node.InPorts.Count)
-                                .Where(node.HasConnectedInput)
+                                .Where(index=>node.InPorts[index].Connectors.Any())
```
Comment (ikeough) [inline src/DynamoCore/Core/CustomNodeManager.cs]: We can't use the `PortModel.IsConnected` method here because that assumes default values as part of connection. Here we _really_ want to know if the port has any connectors, so we use the more idiomatic `Connectors.Any()`

File: src/DynamoCore/Graph/Nodes/CustomNodes/CustomNodeController.cs [hunk 7]
```diff
@@ -16,23 +17,53 @@ public class CustomNodeController<T> : FunctionCallNodeController<T>
...
         {
-            model.InPortData.Clear();
-
             if (Definition.DisplayParameters == null || Definition.Parameters == null)
                 return;
 
-            var inputs = Definition.DisplayParameters.Zip(Definition.Parameters, (dp, p) => Tuple.Create(dp, p.Description, p.DefaultValue));
-            foreach (var p in inputs)
-                model.InPortData.Add(new PortData(p.Item1, p.Item2, p.Item3));
+            var inputs = Definition.DisplayParameters.Zip(Definition.Parameters, (dp, p) => Tuple.Create(dp, p.Description, p.DefaultValue)).ToList();
```
Comment (ikeough) [inline src/DynamoCore/Graph/Nodes/CustomNodes/CustomNodeController.cs]: The following logic originally lived on `NodeModel` in `RegisterOutputs()`. We move it here for custom nodes so that their ports are not deleted when the underlying function definition changes, but the ports are updated.

File: src/DynamoCore/Graph/Nodes/NodeModel.cs [hunk 8]
```diff
@@ -1281,28 +1284,28 @@ public void SelectNeighbors()
         /// Sets the <seealso cref="ElementState"/> for the node based on
         /// the port's default value status and connectivity.
         /// </summary>
-        public void ValidateConnections()
```
Comment (ikeough) [inline src/DynamoCore/Graph/Nodes/NodeModel.cs]: API Change: `NodeModel.ValidateConnections()` has been removed.

File: src/DynamoCore/Graph/Nodes/NodeModel.cs [hunk 9]
```diff
@@ -1387,35 +1390,12 @@ internal double GetPortVerticalOffset(PortModel portModel)
         ///     Reads inputs list and adds ports for each input.
         /// </summary>
         [Obsolete("RegisterInputPorts is deprecated, please use the InPortNamesAttribute, InPortDescriptionsAttribute, and InPortTypesAttribute instead.")]
-        public void RegisterInputPorts()
+        public void RegisterInputPorts(IEnumerable<PortData> portDatas)
```
Comment (ikeough) [inline src/DynamoCore/Graph/Nodes/NodeModel.cs]: API Change: `RegisterInputPorts()` is now `RegisterInputPorts(IEnumerable<PortData> portDatas)`

File: src/DynamoCore/Graph/Nodes/NodeModel.cs [hunk 10]
```diff
@@ -1429,48 +1409,18 @@ public void RegisterInputPorts()
...
-            RaisesModificationEvents = true;
-            OnNodeModified();
         }
 
         /// <summary>
         ///     Reads outputs list and adds ports for each output
         /// </summary>
         [Obsolete("RegisterOutputPorts is deprecated, please use the OutPortNamesAttribute, OutPortDescriptionsAttribute, and OutPortTypesAttribute instead.")]
-        public void RegisterOutputPorts()
+        public void RegisterOutputPorts(IEnumerable<PortData> portDatas)
```
Comment (ikeough) [inline src/DynamoCore/Graph/Nodes/NodeModel.cs]: API Change: `NodeModel.RegisterOutputPorts()` is now `NodeModel.RegisterOutputPorts(IEnumerable<PortData> portDatas)`

File: src/DynamoCore/Graph/Nodes/NodeModel.cs [hunk 11]
```diff
@@ -1603,7 +1559,7 @@ public void RegisterAllPorts()
         /// <param name="data"></param>
         /// <param name="index"></param>
         /// <returns></returns>
-        public PortModel AddPort(PortType portType, PortData data, int index)
+        private PortModel AddPort(PortType portType, PortData data, int index)
```
Comment (ikeough) [inline src/DynamoCore/Graph/Nodes/NodeModel.cs]: API Change: `NodeModel.AddPort(...)` has been removed. Ports can now be added using `NodeModel.InPorts.Add(...)`

File: src/DynamoCore/Graph/Nodes/PortModel.cs [hunk 12]
```diff
@@ -55,21 +48,27 @@ public ObservableCollection<ConnectorModel> Connectors
...
+            get;
+            internal set;
         }
 
         /// <summary>
         /// Tooltip of the port.
         /// </summary>
         [JsonProperty("Description")]
-        public string ToolTipContent
+        public string ToolTip
```
Comment (ikeough) [inline src/DynamoCore/Graph/Nodes/PortModel.cs]: API Change: The `PortModel.ToolTipContent` property is now `PortModel.ToolTip`.

File: src/DynamoCore/Graph/Nodes/PortModel.cs [hunk 13]
```diff
@@ -180,26 +169,18 @@ public bool UsingDefaultValue
...
-                RaisePropertyChanged("ToolTipContent");
+                RaisePropertyChanged("ToolTip");
             }
         }
 
         /// <summary>
-        /// Controls whether the Use Default Value option is available.
-        /// </summary>
-        [JsonIgnore]
-        public bool DefaultValueEnabled
```
Comment (ikeough) [inline src/DynamoCore/Graph/Nodes/PortModel.cs]: API Change: `PortModel.DefaultValueEnabled` has been removed. Please use `PortModel.DefaultValue == null` to check whether a port has a default value.

File: src/DynamoCore/Graph/Nodes/PortModel.cs [hunk 14]
```diff
@@ -282,24 +263,34 @@ public bool ShouldKeepListStructure
             }
         }
 
+        /// <summary>
+        /// Returns true if the port has connectors or if the 
+        /// default value is enabled and not null. Otherwise, returns false.
+        /// </summary>
+        internal bool IsConnected
```
Comment (ikeough) [inline src/DynamoCore/Graph/Nodes/PortModel.cs]: API Change: `PortModel.IsConnected` property is added. This property replaces functionality previously contained in the `NodeModel.HasInput` and `NodeModel.HasConnectedInput` properties.

File: src/DynamoCore/Graph/Nodes/PortModel.cs [hunk 15]
```diff
@@ -109,16 +108,6 @@ public int Index
...
         {
-            get { return portData.LineIndex; }
-        }
-
-        /// <summary>
-        /// A flag indicating whether the port is considered connected.
-        /// </summary>
-        [Obsolete("Please use NodeModel.HasConnectedInput instead.")]
-        [JsonIgnore]
-        public bool IsConnected
```
Comment (ikeough) [inline src/DynamoCore/Graph/Nodes/PortModel.cs]: API Change: Obsolete `PortModel.IsConnected` property has been removed.

File: src/DynamoCoreWpf/Commands/NodeCommands.cs [hunk 16]
```diff
@@ -106,18 +106,6 @@ public DelegateCommand ViewCustomNodeWorkspaceCommand
             }
         }
 
-        public DelegateCommand ValidateConnectionsCommand
```
Comment (ikeough) [inline src/DynamoCoreWpf/Commands/NodeCommands.cs]: API Change: `NodeCommands.ValidateConnectionsCommand` has been removed.

**Comment by ikeough** [comment]
@mjkkirschner ping. Ready for another round. Call me to chat if you want to know more about some of these refactors.

File: src/DynamoCore/Graph/Nodes/PortModel.cs (hunk 13 above)
Comment (mjkkirschner) [inline src/DynamoCore/Graph/Nodes/PortModel.cs]: I've got one concern here and this is a lot of zt node authors desire to actually use null as a default value...  maybe that design decision is out of scope for this PR and this won't actually change the behavior (I dont think it will) just wanted to mention it.

File: src/DynamoCoreWpf/Views/Core/NodeView.xaml.cs [hunk 17]
```diff
@@ -366,17 +366,6 @@ internal void DisableInteraction()
                 ViewModel.SetStateCommand.Execute(ElementState.Dead);
         }
 
-        internal void EnableInteraction()
-        {
-            foreach (UIElement e in inputGrid.Children)
-            {
-                if (enabledDict.ContainsKey(e))
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Views/Core/NodeView.xaml.cs]: did anything use this dictionary?

**Comment by mjkkirschner** [comment]
@ikeough this looks pretty good to me.

**Comment by mjkkirschner** [comment]
:shipit:

**Comment by mjkkirschner** [comment]
because this change will effect our internal libraries and many external ones, should we create an issue to track public documentation / update notes on this change and what internal repos still need to be updated?

**Comment by ikeough** [comment]
@mjkkirschner I'll make a project for it :)

**Comment by ikeough** [comment]
Project added: https://github.com/DynamoDS/Dynamo/projects/3. I'll tackle these in short order.
